### ISSUE TITLE
Using recorder to only track specific information?

### ISSUE THREAD

**Issue by johnabs** [MilesCranmer/SymbolicRegression.jl#55]
Hi there,

I'm using recorder (which doesn't actually show up in the documentation anywhere that I've seen by the way, I found it in a github issue). But instead of looking at all the information, I just need a small portion: specifically loss values. (I'm looking at quantitatively representing convergence rate in terms of loss vs iterations). Additionally, I noticed there seem to be both "population" entries in the JSON file, as well as what I can only refer to as "individual" values? Basically the "individuals" have keys that are a large string of numbers, and they contain values such as events, score, tree, parent, etc. But in the populations, there's much simpler information per entry. 

Basically, I just need to know 2 things:

1. Is there a way to filter what is stored in the output with recorder so I'm not left with 100s of MB files that I have to parse?
2. Which set of information here, ("individual" vs entrys in the population dictionaries) should I use to check convergence rate? Or are these two identical, but with different levels of detail?

Thanks for your help!

Best,

John

**Comment by MilesCranmer** [comment]
Hi John,

I created the recorder for internal debugging, not so much for use by others, hence why I chose not to document. But it sounds like it is useful for some so I will eventually clean it up and then add docs.

So, IIRC, the population statistics are recorded once every iteration. Each iteration contains many many mutations of individual equations. The `individual` entries are meant to track every single mutation. The keys shown are meant to track the "family tree" of a particular equation, which is why they give so much detail - there are quite a few mutations and these are all stored.

1. There is no flag to do this. I made it so that all information would be stored, and then I could filter down the JSON afterwards. This sounds like it's not optimal for what you would like, so you could try tweaking the code. 

These two lines:
https://github.com/MilesCranmer/SymbolicRegression.jl/blob/2e4edced6f025f9d3aee59ae7d330977e8dc2962/src/RegularizedEvolution.jl#L79-L80
are what stores the mutation information in the recorder struct. I think if you turn these off, it should work. I can't help more at this stage due to a crazy schedule for the coming month but in the future I can look at this.

2. This is a very complex question to ask for genetic algorithms, so not sure if there is a clear answer. There are many different things you could look at: the convergence of a particular equation, or the convergence of a population of equations to some best equation. One thing you could try first is just looking at the best equations recorded each iteration and check the convergence of that. But that would miss convergence of individual populations... Again, it's hard to ask things like this for GAs. For regular ML it makes sense, but GAs search combinatorial spaces so it is a bit trickier.
Cheers,
Miles

**Comment by johnabs** [comment]
Hi Miles,

Thank you for pointing me in the right direction here; I'll take a look at the code and possibly create a PR with some changes to allow pre-filtering the output if you think that would be helpful :)

Thanks again for all your help!

Best,

John

**Comment by johnabs** [comment]
Hi Miles,

As a quick followup, I'm struggling to find which source file contains the code that exports the json files. I want to get rid of the mutations and options fields, to just keep the populations fields (which I've already filtered down to just loss and tree). I think I checked all the files for this, but maybe it's hiding somewhere that I didn't think to look.

Best,

John

**Comment by MilesCranmer** [comment]
No worries, here is where the current population is recorded:
https://github.com/MilesCranmer/SymbolicRegression.jl/blob/d107209410887c6e91d5374e0f7118c7e38a818b/src/SymbolicRegression.jl#L475.

The `@recorder` macro simply checks if recording is turned on, using `options`. I think you could turn recording off, then hardcode every `@recorder` line to be always true in the parts that you want? You will need this:
https://github.com/MilesCranmer/SymbolicRegression.jl/blob/d107209410887c6e91d5374e0f7118c7e38a818b/src/SymbolicRegression.jl#L468-L471
so that it stores the correct iteration. 

As well as this:
https://github.com/MilesCranmer/SymbolicRegression.jl/blob/d107209410887c6e91d5374e0f7118c7e38a818b/src/SymbolicRegression.jl#L325
which stores the first iteration.

Then, here is where the JSON is written:
https://github.com/MilesCranmer/SymbolicRegression.jl/blob/d107209410887c6e91d5374e0f7118c7e38a818b/src/SymbolicRegression.jl#L570-L574

**Comment by johnabs** [comment]
Awesome, I also have one (hopefully final) question. Is there a reason the JSON file seems to write the keys non-sequentially?

It looks like the order is pop2, then 19, then 7, etc. rather than being in chronological order. Even stranger, I'm not multi-threading this (at least that I'm aware of, as I've set numprocs to 1).
